### PR TITLE
UsabilityHub Topping Exporter

### PR DESCRIPTION
**Pull request by signedav**
Wizard to create toppings from an existing QGIS Project .

[Check the screencast](https://user-images.githubusercontent.com/28384354/186889072-e9af3e58-b05f-4a64-b257-e868e5838e8a.webm)

### Concept
Concept can be found here: https://gist.github.com/signedav/44b12b6667c707f646dc6177e026ac56#file-baker_usabilityhub_exporter_final-md

#### Frontend
In the folder `topping_wizard` all the GUI components are. The information about the project are loaded from the `QgsProject.instance()` and from the layers sources (connecting the database over the `modelbaker` library). The `ToppingWizard` contains an object of the `IliProjectTopping` that stores all the settings concerning the topping and provides the functionality to create the topping files.

#### Backend
The backend contains two "packages". They are stored in `internal_libs` but should be moved to another place (not yet sure where to).
##### toppingmaker
Contains `ProjectTopping`. This object contains a layertree-configuration and the `ExportSettings` containing the info, how the layers should be exported (like as QML file etc.). The `Target` object contains the target paths and a path_resolver creating the links to the exported topping files and collecting all the created topping files. The result is a YAML containing layertree and layerorder and all the exported topping files of the layers (QML, QLR).
See https://github.com/opengisch/toppingmaker/pull/1

##### ilitoppingmaker
Contains `IliProjectTopping` derived from `ProjectTopping`, where additional parameters are used for a UsabILIty Hub INTERLIS specific project. This means it contains an object of `MetaConfig` (resulting in an INI file), `IliTarget` (derived from `Target`, containing a UsabILIty Hub specific path resolver (resulting in ilidata:-Ids)) and connects a `IliData` class to create the ilidata.xml according to all the listed topping files in the `IliTarget`.
See https://github.com/opengisch/QgisModelBakerLibrary/pull/23

Depends on https://github.com/opengisch/QgisModelBakerLibrary/pull/18

### Not parts of the implementation
- In the layer selection there could be some more logic (like when selecting a QLR definition export on a group the contained layers are removed from the tree (etc).
- The ili2db Parameters loaded from the Schema cannot be edited (except the pre- and postscript and the meta attribute file (the-file-formerly-known-as-toml). But anyway, the settings from the schema are written to the ini, but this part is not yet supported by ili2db. See https://github.com/claeis/ili2db/issues/392
- A workflow to merge the ilidata.xml with an existing one (not part of this project). Maybe with ili2db - maybe with other functionality. But not scope of this implementation.
- in `toppingmaker` functions like `loadFiles` and `generateProject` to make the other way around and create the QgsProject..

##### Financed
by a freundliche Colaboration der Kantonen Schaffhausen, Appenzell Innerhoden und Schwyz

### CODE REVIEW

**Comment by signedav** [comment]
@domi4484 Before you have a look at this :mammoth: PR I suggest we will have a call where we go through.

@m-kuhn Requested your review mainly from the part in `/internal_libs/toppingmaker` and to have your opinion where to put this part (and the `ilitoppingmaker` part as well). While the `ilitoppingmaker` could be in the `modelbaker` library I'm not sure if we should package `toppingmaker` as an own library provided at PYPI. What do you think?

**Review by domi4484 — APPROVED** [review]
sorry i don't have much inputs, but overall code looks good!

File: QgisModelBaker/gui/topping_wizard/referencedata_page.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        completer = QCompleter(
+            self.ilireferencedatacache.model,
+            self.input_line_edit,
+        )
+        completer.setCaseSensitivity(Qt.CaseInsensitive)
+        completer.setFilterMode(Qt.MatchContains)
+        completer.popup().setItemDelegate(self.ilireferencedata_delegate)
+        self.input_line_edit.setCompleter(completer)
+        self.input_line_edit.textChanged.emit(self.input_line_edit.text())
+        self.input_line_edit.textChanged.connect(self._complete_referencedata_completer)
```
[the remark this answers is not in the document]
Comment (signedav) [reply QgisModelBaker/gui/topping_wizard/referencedata_page.py:93]: Oh. Hm... I'll check that.

File: QgisModelBaker/gui/topping_wizard/target_page.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+├─ {subfolder}/
+    ├─ metaconfig/
+    │  ├─ {projectname_slug}.ini
+    ├─ qml/
+    │  ├─ {projectname_slug}_name_of_the_layer.qml
+    │  ├─ {projectname_slug}_name_of_another_layer.qml
+    ├─ sql/
+    │  ├─ the_script.sql
+    ├─ etc.
+            """
```
[the remark this answers is not in the document]
Comment (signedav) [reply QgisModelBaker/gui/topping_wizard/target_page.py:144]: Copied from my notes I made back in January. Don't remember. But you can do it in Ubuntu with `tree`

File: QgisModelBaker/internal_libs/ilitoppingmaker/__init__.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ ***************************************************************************/
+"""
+
+from QgisModelBaker.internal_libs.toppingmaker import ExportSettings
+
+from .ili2dbsettings import Ili2dbSettings
+from .ilidata import IliData
+from .iliprojecttopping import IliProjectTopping
+from .ilitarget import IliTarget
+from .metaconfig import MetaConfig
```
[the remark this answers is not in the document]
Comment (signedav) [reply QgisModelBaker/internal_libs/ilitoppingmaker/__init__.py:27]: Check out this comment https://stackoverflow.com/questions/10863268/how-is-an-empty-init-py-file-correct#10863282

[the remark this answers is not in the document]
Comment (domi4484) [reply QgisModelBaker/internal_libs/ilitoppingmaker/__init__.py:27]: Thanks!

**Pull request merged**